Merges mob properties with traits, adding value traits & modularising mutant skin colour

PowerfulBacon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

fixes #11759

## About The Pull Request

Merges mob properties and traits, for a convenient method of getting values from a trait.

```dm
// Set the skin colour trait value to 0, with a priority of 50
ADD_TRAIT(target, TRAIT_SKIN_COLOUR, TRAIT_SOURCE_HUSK, "#000", 50)
// '#000'
GET_TRAIT_VALUE(target, TRAIT_SKIN_COLOUR)
// 'true'
HAS_TRAIT(target, TRAIT_SKIN_COLOUR)
// Add a new skin colour with higher priority
ADD_TRAIT(target, TRAIT_SKIN_COLOUR, TRAIT_SOURCE_RED, "#f00", 51)
// '#f00'
GET_TRAIT_VALUE(target, TRAIT_SKIN_COLOUR)
// Traits can be added like normal too
ADD_TRAIT(target, TRAIT_HAPPY, SOURCE_DEPRESSION)
```

Note: This replaces the lists used internally by traits with a heap.

Also adds unit tests for traits, considering how important they are.

As a secondary effect, implements a TODO which was to optimise pathing by removing some proc call overhead. This will make a negligable difference, not sure why the coder thought it was that important to begin with tbh.

All trait types share the same methods to remove, and test:
- HAS_TRAIT
- REMOVE_TRAIT
- REMOVE_TRAIT_NOT_FROM
- REMOVE_TRAITS_NOT_IN
- REMOVE_TRAITS_IN

## Traits

Traits are a boolean status on a mob. A mob either has a trait or it doesn't.

A trait can come from multiple sources, making it useful for having modular code that wants to do the same thing without conflicting with other modules.

```dm
ADD_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE)
REMOVE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE)
```

## Value Traits

Value traits have a priority and a value.

When the value is fetched, the value with the highest priority is the value returned.

This is useful for modules wanting to set the same variable and having them deconflict based on priority

```dm
var/value = new /datum() // Any value is valid
ADD_VALUE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE, value, 10)
REMOVE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE)
var/result = GET_TRAIT_VALUE(target, TRAIT_EXAMPLE)
```

## Numeric Traits

Numeric traits are a combination of additors and multipliers.

The final value is calculated by taking the sum of all the add values, and then applying the multipliers.

This allows multiple modules to modify a number with the deconflict method applying all operations in a deterministic order. This is useful if you have a stat that can increase by 50% or by +5.

```dm
// Add 5
ADD_CUMULATIVE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE_1, 5)
// Multiply by 0.5
ADD_MULTIPLICATIVE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE_2, 0.5)
// Returns 2.5
var/result = GET_VALUE_TRAIT(target, TRAIT_EXAMPLE)
// Remove the multiplier
REMOVE_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE_2)
// Returns 5
result = GET_VALUE_TRAIT(target, TRAIT_EXAMPLE)
```

## Overlapping

Adding the same trait from the same source will overlap the traits and will not replace the old value. This should be avoided, but the defined behaviour is to allow stacking.

For example:
```dm
ADD_NUMERIC_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE_1, 5)
ADD_NUMERIC_TRAIT(target, TRAIT_EXAMPLE, SOURCE_EXAMPLE_1, 5)
// Get trait would produce 10, not 5
```

## Why It's Good For The Game

Mob properties are abandoned since they are a weird duplicate of traits. This merges them together, and allows values to be put into traits in a convenient manner.

We are all used to traits, and nobody uses properties. We might as well not have 2 systems that do the same thing and make something that we can expand upon but remains compatible with TG. Mob properties had a lot of room for improvements, so here they are.

We can commence forth the mass-deshitcodification where we remove all of the random specific variables that store something thats rarely used and instead merge them into traits.

Why skin colour?
- I plan to make the nanite armour buff an active ability which turns your skin grey, and I don't like adding in random specific checks for nanites slap bang in the middle of mob code.

## Testing Photographs and Procedure

![image](https://github.com/user-attachments/assets/e87f885d-80c3-40bd-828a-71440fd25b61)

![image](https://github.com/user-attachments/assets/fd6e578d-47fe-442c-b3df-982f4b74a0f3)

![image](https://github.com/user-attachments/assets/2048faad-4e1a-4eb3-a800-15866b7135be)

Bodyparts dropped correctly have no colour.

## Performance

There are some minor performance regressions, however I have worked to minimise the effects of these as much as possible.

Add trait:

![image](https://github.com/user-attachments/assets/b2a55083-71e2-47b7-a0e2-a4378768a8d4)

Edit: After some refactoring, ADD_TRAIT will now perform no different to before as the different types of traits have been split up.

Remove trait:

![image](https://github.com/user-attachments/assets/17fb8c18-89a0-43fc-ac80-bd3ff31a90fe)

Has trait:
![image](https://github.com/user-attachments/assets/d7b279bd-497e-46a5-9e4f-333e7dfe3f4a)


Has trait in:
- Will have some proc call overhead

## Changelog
:cl:
code: Merges mob properties and traits.
fix: Hulk limb colour no longer applies to robotic limbs.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PowerfulBacon committed:
Merges mob properties

Update trait_tests.dm

Update trait_tests.dm

Deletes properties

Fixes some bugs with traits and heaps

Update trait_tests.dm

Get heap values

Fixes a bug caused by a byond issue

Update _traits.dm

priority trait value test

Adds more trait tests

Adds support for trait not in

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Fixes remove in and remove not in

Mutant bodyparts

Update _traits.dm

Update trait_tests.dm

Update _traits.dm

Update _traits.dm

Absolute path conversion

Co-Authored-By: EvilDragonfiend <87972842+EvilDragonfiend@users.noreply.github.com>

Adds the traits to the trait declaration file

Update heap.dm

Heavy optimisations

Update heap.dm

Update heap.dm

Further performance improvements

Fixes traits for non string sources

Update _traits.dm

Update _traits.dm

Spelling mistake fix

Removes linting for the static control flow term

Fixes some linting issues

Update heap.dm

Adds open dream trait support

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update heap.dm

Update _traits.dm

Update _traits.dm

Tries to fix opendream

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Update _traits.dm

Seperates value traits and normal traits to appease OD

Update trait_tests.dm

Update _traits.dm

Fixes a mix up with variable naming

Cleans up the code

EvilDragonfiend commented on code/__DEFINES/_heap_helpers/heap.dm:
```diff
@@ -0,0 +1,95 @@
+			sink(index);\
+		}\
+	}\
+}
```
```suggestion
#define HEAP_TYPE(typepath, compare_value) \
##typepath{\
	var/list/L;\
}\
##typepath/New(...) {\
	L = list(args);\
}\
##typepath/Destroy(force, ...) {\
	for(var/i in L) {\
		qdel(i);\
	}\
	L = null;\
	return ..();\
}\
##typepath/proc/is_empty() {\
	return !length(L);\
}\
##typepath/proc/insert(atom/A) {\
	L.Add(A);\
	swim(length(L));\
}\
##typepath/proc/pop() {\
	if(!length(L)) {\
		return 0;\
	}\
	. = L[1];\
	L[1] = L[length(L)];\
	L.Cut(length(L));\
	if(length(L)) {\
		sink(1);\
	}\
}\
##typepath/proc/swim(index) {\
	var/parent = round(index * 0.5);\
	while(parent > 0 && (L[index]:##compare_value - L[parent]:##compare_value > 0)) {\
		L.Swap(index,parent);\
		index = parent;\
		parent = round(index * 0.5);\
	}\
}\
##typepath/proc/sink(index) {\
	var/g_child = get_greater_child(index);\
	while(g_child > 0 && (L[index]:##compare_value - L[g_child]:##compare_value < 0)) {\
		L.Swap(index,g_child);\
		index = g_child;\
		g_child = get_greater_child(index);\
	}\
}\
##typepath/proc/get_greater_child(index) {\
	if(index * 2 > length(L)) {\
		return 0;\
	}\
	if(index * 2 + 1 > length(L)) {\
		return index * 2;\
	}\
	if(L[index * 2]:##compare_value - L[index * 2 + 1]:##compare_value < 0) {\
		return index * 2 + 1;\
	} else {\
		return index * 2;\
	}\
}\
##typepath/proc/resort(atom/A) {\
	var/index = L.Find(A);\
	swim(index);\
	sink(index);\
}\
##typepath/proc/List() {\
	. = L.Copy();\
}\
##typepath/proc/operator+=(A) {\
	L.Add(A);\
	swim(length(L));\
}\
##typepath/proc/operator|=(A) {\
	var/original_length = length(L);\
	L |= A;\
	if (original_length != length(L)) {\
		swim(length(L));\
	}\
}\
##typepath/proc/operator-=(A) {\
	var/index = L.Find(A);\
	if(index == 0) {\
		return 0;\
	}\
	. = L[index];\
	L[index] = L[length(L)];\
	L.Cut(length(L));\
	if(length(L)) {\
		sink(index);\
	}\
}
```
This will work

PowerfulBacon replied:
This was how it was originally but didn't work due to comments (which I removed anyway). Been meaning to convert it back, but not sure how much byond actually liked this.

PowerfulBacon committed:
Merge branch 'master' into merges-mob-properties-and-traits

PowerfulBacon committed:
09/11/2024 Code cleanup + Numeric Trait Expansion

Cleaning the code part 1

Update _traits.dm

Fixes failing tests

Update _traits.dm

Update _traits.dm

Update _traits.dm

Adds multiplicative and cumulative traits

Fixes compilation issue with numeric traits

Adds in the ability to remove cum values

Fixes a bug in the trait tests which resulted in incorrect logic

Full recalculation to deal with precision loss

Update _traits.dm

Fixes incorrect indentation...

Update _traits.dm

Fixes the trait tests using the wrong form of remove trait, and thus concludes trait work

PowerfulBacon committed:
Fixes a compilation time issue

Update trait_tests.dm

Fixes a bug caught by unit testing

Fixes a bug with defines that happened to pass the unit tests

And again

Adds in a mode for debug compilation, which speeds things up

Update _traits.dm

Removes some unnecessary complexity

Fixes compilation times

PowerfulBacon committed:
Fixes an erronously removed remove_heap

PowerfulBacon committed:
Merge branch 'master' into merges-mob-properties-and-traits

EvilDragonfiend reviewed (changes requested):
Would you please put * in parameter explanations? VSCode doesn't recognise newline well when you see the proc hint tooltip.
Also, instead of numbering, specifying parameter name will be easier to read.
I am not sure if you'd like my style, but I'd write in this way:
```
/// Add a trait to a target
/// Parameters:
/// 1: The target to receive the trait
/// 2: The key of the trait
/// 3: The source of the trait
/// 4: The priority of the trait value
/// 5: The value stored in the trait
``` 
to 
```
/// Add a trait to a target
/// * target<datum>: The target to receive the trait
/// * trait_type<DEFINE>: The key of the trait
/// * source<datum|text>: The source of the trait
/// * trait_value<datum>: The value stored in the trait
/// * trait_priority<int>: The priority of the trait value
```
(also, you miswrote 4 and 5. this would be the reason why specifying name is better, I suppose.)

-------------

Also, I feel like there should be a readme that describes the macros.
The PR body seems to be good to be the readme.
(but please update PR body. it's outdated. there is no `MULTIPLY_NUMERIC_TRAIT`)

because I feel overwhelemed massive size of the dm file, and I have no idea which proc is what I would need when especially these are all quite similar.

------------

This is a bit confusing to understand yet, but I see this is ultra-potential.
I think ths confusion comes from the absense of the correct document.
and I will think about how this can be flowing to read apart from the confusion. (I feel trait datum and macros are a-bit-not connected well to read to each others.)

PowerfulBacon commented on code/__DEFINES/traits/_traits.dm:
```diff
@@ -1,119 +1,339 @@
+/// 2: The key of the trait
+/// 3: The source of the trait
+/// 4: The priority of the trait value
+/// 5: The value stored in the trait
```
```suggestion
/// 4: The value stored in the trait
/// 5: The priority of the trait value
```

PowerfulBacon committed:
Update code/__DEFINES/traits/_traits.dm

PowerfulBacon committed:
Merge remote-tracking branch 'upstream/master' into merges-mob-properties-and-traits

PowerfulBacon committed:
Update _traits.dm

PowerfulBacon committed:
Update _traits.dm
